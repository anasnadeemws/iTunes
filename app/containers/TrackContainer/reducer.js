/*
 *
 * TrackContainer reducer
 *
 */
import produce from 'immer';
import { createActions } from 'reduxsauce';
import get from 'lodash/get';

export const { Types: trackContainerTypes, Creators: trackContainerCreators } = createActions({
  requestGetGithubRepos: ['repoName'],
  successGetGithubRepos: ['data'],
  failureGetGithubRepos: ['error'],
  clearGithubRepos: {}
});
export const initialState = { repoName: null, reposData: {}, reposError: null };

/* eslint-disable default-case, no-param-reassign */
export const trackContainerReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case trackContainerTypes.REQUEST_GET_GITHUB_REPOS:
        draft.repoName = action.repoName;
        break;
      case trackContainerTypes.CLEAR_GITHUB_REPOS:
        draft.repoName = null;
        draft.reposError = null;
        draft.reposData = {};
        break;
      case trackContainerTypes.SUCCESS_GET_GITHUB_REPOS:
        draft.reposData = action.data;
        break;
      case trackContainerTypes.FAILURE_GET_GITHUB_REPOS:
        draft.reposError = get(action.error, 'message', 'something_went_wrong');
        break;
    }
  });

export default trackContainerReducer;
