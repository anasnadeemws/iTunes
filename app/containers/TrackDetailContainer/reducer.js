/*
 *
 * TrackDetailContainer reducer
 *
 */
import produce from 'immer';
import { createActions } from 'reduxsauce';
import get from 'lodash/get';

export const { Types: trackDetailContainerTypes, Creators: trackDetailContainerCreators } = createActions({
  requestGetTrackDetail: ['trackId'],
  successGetTrackDetail: ['data'],
  failureGetTrackDetail: ['error']
});
export const initialState = { trackId: null, trackDetailData: {}, trackError: null, trackDetailLoading: false };

/* eslint-disable default-case, no-param-reassign */
export const trackDetailContainerReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case trackDetailContainerTypes.REQUEST_GET_TRACK_DETAIL:
        draft.trackId = action.trackId;
        draft.trackDetailLoading = true;
        break;
      case trackDetailContainerTypes.SUCCESS_GET_TRACK_DETAIL:
        draft.trackDetailData = action.data;
        draft.trackDetailLoading = false;
        break;
      case trackDetailContainerTypes.FAILURE_GET_TRACK_DETAIL:
        draft.trackError = get(action.error, 'message', 'something_went_wrong');
        draft.trackDetailLoading = false;
        break;
    }
  });

export default trackDetailContainerReducer;
