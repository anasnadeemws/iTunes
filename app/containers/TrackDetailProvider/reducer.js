/*
 *
 * TrackDetailContainer reducer
 *
 */
import produce from 'immer';
import { createActions } from 'reduxsauce';
import get from 'lodash/get';

export const { Types: trackDetailTypes, Creators: trackDetailCreators } = createActions({
  requestGetTrackDetail: ['trackId'],
  successGetTrackDetail: ['data'],
  failureGetTrackDetail: ['error']
});
export const initialState = { trackId: null, trackDetailData: {}, trackError: null, trackDetailLoading: false };

/* eslint-disable default-case, no-param-reassign */
export const trackDetailReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case trackDetailTypes.REQUEST_GET_TRACK_DETAIL:
        draft.trackId = action.trackId;
        draft.trackDetailLoading = true;
        break;
      case trackDetailTypes.SUCCESS_GET_TRACK_DETAIL:
        draft.trackDetailData = action.data;
        draft.trackDetailLoading = false;
        break;
      case trackDetailTypes.FAILURE_GET_TRACK_DETAIL:
        draft.trackError = get(action.error, 'message', 'something_went_wrong');
        draft.trackDetailLoading = false;
        break;
    }
  });

export default trackDetailReducer;
