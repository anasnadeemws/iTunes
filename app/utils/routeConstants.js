export default {
  repos: {
    route: '/',
    props: {
      maxwidth: 500,
      padding: 20
    },
    exact: true
  },
  tracks: {
    route: '/tracks',
    props: {
      maxwidth: 500,
      padding: 20
    },
    exact: true
  },
  track: {
    route: '/track:trackid',
    props: {
      maxwidth: 500,
      padding: 20
    },
    exact: true
  }
};
