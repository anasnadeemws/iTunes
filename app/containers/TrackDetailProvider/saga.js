import { put, call, takeLatest } from 'redux-saga/effects';
import { getTrack } from '@services/trackApi';
import { trackDetailTypes, trackDetailCreators } from './reducer';

const { REQUEST_GET_TRACK_DETAIL } = trackDetailTypes;
const { successGetTrackDetail, failureGetTrackDetail } = trackDetailCreators;
export function* getTrackDetail(action) {
  const response = yield call(getTrack, action.trackId);
  const { data, ok } = response;
  if (ok) {
    yield put(successGetTrackDetail(data));
  } else {
    yield put(failureGetTrackDetail(data));
  }
}
// Individual exports for testing
export default function* trackDetailSaga() {
  yield takeLatest(REQUEST_GET_TRACK_DETAIL, getTrackDetail);
}
