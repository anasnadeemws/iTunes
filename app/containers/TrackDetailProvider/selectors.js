import { createSelector } from 'reselect';
import get from 'lodash/get';
import { initialState } from './reducer';

/**
 * Direct selector to the trackDetailContainer state domain
 */

export const selectTrackDetailDomain = (state) => state.trackDetailContainer || initialState;

/**
 * Other specific selectors
 */

/**
 * Default selector used by trackDetailContainer
 */

export const selectTrackDetailData = () =>
  createSelector(selectTrackDetailDomain, (substate) => get(substate, 'trackDetailData'));

export const selectTrackError = () =>
  createSelector(selectTrackDetailDomain, (substate) => get(substate, 'trackError'));

export const selectTrackDetailLoading = () =>
  createSelector(selectTrackDetailDomain, (substate) => get(substate, 'trackDetailLoading'));
