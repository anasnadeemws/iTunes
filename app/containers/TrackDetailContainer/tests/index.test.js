/**
 *
 * Tests for TrackDetailContainer
 *
 */

import React from 'react';
import { renderProvider } from '@utils/testUtils';
import { TrackDetailContainerTest as TrackDetailContainer, mapDispatchToProps } from '../index';
import { translate } from '@app/utils/index';
import { trackDetailTypes } from '@app/containers/TrackDetailProvider/reducer';

describe('<TrackDetailContainer /> tests', () => {
  let submitSpy;

  beforeEach(() => {
    submitSpy = jest.fn();
  });
  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(<TrackDetailContainer dispatchTrackDetail={submitSpy} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should validate mapDispatchToProps actions', async () => {
    const dispatchTracksSearchSpy = jest.fn();
    const trackId = 12323;
    const actions = {
      dispatchTrackDetail: { trackId, type: trackDetailTypes.REQUEST_GET_TRACK_DETAIL }
    };

    const props = mapDispatchToProps(dispatchTracksSearchSpy);
    props.dispatchTrackDetail(trackId);
    expect(dispatchTracksSearchSpy).toHaveBeenCalledWith(actions.dispatchTrackDetail);
  });

  it('should render default error message when search goes wrong', () => {
    const defaultError = translate('something_went_wrong');
    const { getByTestId } = renderProvider(
      <TrackDetailContainer trackError={defaultError} dispatchTrackDetail={submitSpy} />
    );
    expect(getByTestId('error-message')).toBeInTheDocument();
    expect(getByTestId('error-message').textContent).toBe(defaultError);
  });
});
