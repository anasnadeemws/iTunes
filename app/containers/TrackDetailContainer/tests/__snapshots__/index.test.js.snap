// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackDetailContainer /> tests should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg css-psxadm-MuiContainer-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root e12nrugs5 css-g118n2-MuiPaper-root-MuiCard-root"
        data-testid="track-card"
      >
        <div
          class="e12nrugs4 MuiBox-root css-1ahrg9s"
        >
          <div
            class="e12nrugs2 MuiBox-root css-yhs9f2"
          >
            <a
              href="/tracks/undefined"
            >
              <div
                class="MuiTypography-root MuiTypography-h5 css-41an0y-MuiTypography-root"
              >
                <p
                  class="css-lh9t18 egtqi0h0"
                  data-testid="collection_name_unavailable"
                >
                  Track name is unavailable
                </p>
              </div>
            </a>
          </div>
          <div
            class="MuiTypography-root MuiTypography-subtitle1 css-ecpxie-MuiTypography-root"
          >
            <p
              class="css-lh9t18 egtqi0h0"
              data-testid="track_artist_name_unavailable"
            >
              Artist name is unavailable
            </p>
          </div>
          <p
            class="css-lh9t18 egtqi0h0"
            data-testid="track_shortdesc_unavailable"
          >
            Description unavaiable
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;
